src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,141 @@
+    ShaderProgram::compile(Type const type, std::string_view const source, bool const fallback) {
+        // compile
+
+        auto id{ glCreateShader(sly::to_underlying(type)) };
+        auto const* c_str = source.data();
+        glShaderSource(id, 1, &c_str, nullptr);
+        glCompileShader(id);
+
+        // check
+        GLint success;
+        glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+        if (not success) {
+            GLchar message[512];
+            glGetShaderInfoLog(id, 512, nullptr, message);
+            spdlog::critical("ERROR::SHADER::{}::COMPILATION_FAILED -> {}", get_name_from_type(type), message);
+
+            if (not fallback) {
+                if (type != Type::Geometry) {
+                    glDeleteShader(id);
+                    spdlog::error("ERROR::SHADER::{}::FALLBACK", get_name_from_type(type));
+                    if (type == Type::Vertex) {
+                        return compile(type, fallback_vertex_source, true);
+                    } else { // Type::Fragment
+                        return compile(type, fallback_fragment_source, true);
+                    }
+                } else {
+                    spdlog::critical("ERRPR::SHADER::{}::NO_FALLBACK", get_name_from_type(type));
+                }
+            }
+
+        } else {
+            spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_type(type));
+        }
+
+        if (success) {
+            return { id };
+        } else {
+            glDeleteShader(id);
+            return {};
+        }
+    }
+
+    void ShaderProgram::attach_shader(Type const type, Shader const shader) const {
+        if (not shader.has_value()) {
+            spdlog::error("shader {} not add to program", get_name_from_type(type));
+            return;
+        }
+
+        glAttachShader(m_program_id, shader.value());
+        glDeleteShader(shader.value());
+    }
+
+    void ShaderProgram::link_program() const {
+        glLinkProgram(m_program_id);
+
+        GLint success;
+        glGetProgramiv(m_program_id, GL_LINK_STATUS, &success);
+        if (not success) {
+            GLchar message[512];
+            glGetProgramInfoLog(m_program_id, 512, NULL, message);
+            spdlog::critical("ERROR::PROGRAMM::LINK_FAILED -> {}\n", message);
+        }
+    }
+
+    ShaderProgram::ShaderProgram(
+            std::string_view const vertex_source,
+            std::string_view const geometry_source,
+            std::string_view const fragment_source
+    ) {
+        m_program_id = { glCreateProgram() };
+
+        auto const vertex_shader = compile(Type::Vertex, vertex_source);
+        auto const geometry_shader = compile(Type::Geometry, geometry_source);
+        auto const fragment_shader = compile(Type::Fragment, fragment_source);
+
+        attach_shader(Type::Vertex, vertex_shader);
+        attach_shader(Type::Geometry, geometry_shader);
+        attach_shader(Type::Fragment, fragment_shader);
+
+        link_program();
```

Comment:
Another reason why we should handle the error here: If compiling the shaders fails, linking will also fail (if fallback is disabled). In that case we must throw an exception.
Maybe we should ***always*** use the fallback when there's an error. I don't think it's useful to handle it with a crash or in any other way.

Reply:
what do you mean by `always`?
the geometry shader does not have a fallback because opengl can run without it.
vertex and fragment shader are a must. so they have a fallback.
i think it should be fine when we just dont bind the failed geometry shader.

Reply:
Yes, that's what I mean 👍 I was just referencing the `fallback` parameter of the compile-function. I don't think we need that as a parameter, just always try to fallback if necessary (ignore the geometry shader if necessary).